mountpoint-s3/src/fs.rs

```diff
@@ -604,15 +612,31 @@ where
         atime: Option<OffsetDateTime>,
         mtime: Option<OffsetDateTime>,
         _flags: Option<u32>,
+        size: Option<u64>,
     ) -> Result<Attr, Error> {
         tracing::info!(
-            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?}",
+            "fs:setattr with ino {:?} flags {:?} atime {:?} mtime {:?} size {:?}",
             ino,
             _flags,
             atime,
-            mtime
+            mtime,
+            size
         );
-        let lookup = self.superblock.setattr(&self.client, ino, atime, mtime).await?;
+        let lookup = match (self.superblock.setattr(&self.client, ino, atime, mtime).await, size) {
+            (Ok(lookup), _) => lookup,
+            // For file overwrite, setattr is called with its size field having Some value.
+            // but for other cases like modifying atime, mtime etc the size field have None value.
+            // So, using that as distinguishing factor for both the errors.
+            (Err(e), Some(_)) => {
+                match e {
+                    InodeError::SetAttrNotPermittedOnRemoteInode(_) => return Err(
+                        err!(libc::EPERM, "file overwrite is disabled by default, you need to remount with --allow-overwrite flag and open the file in truncate mode (O_TRUNC) to overwrite it"),
+                    ),
+                    _ => return Err(e.into()),
+                }
```

Comment:
You can probably put `InodeError::SetAttrNotPermittedOnRemoteInode` into the pattern match arm, and then let it fall through to the next arm if it doesn't match.